### DISCUSSION TITLE
역량 / 리포트 관련 시나리오

### DISCUSSION THREAD

**Discussion by gracefulBrown** [woowacourse/prolog#610] [General]
# 도메인 설명
## 역량
- 사용자의 역량 정보
- 기존에 작성한 학습로그를 역량과 맵핑하여 역량에 대한 증빙을 할 수 있음
- 역량(과 역량-학습로그 맵핑)은 수정할 수 있으며 수정할 때 마다 역량 이력이 생성됨(커밋처럼)

## 역량 이력
- 역량의 변경 히스토리 정보

## 리포트
- 역량(이력)의 추가적인 설명 정보
- 리포트에는 제목과 설명을 입력할 수 있고 역량에 대한 가중치를 설정할 수 있음

# 역량
- 역량을 설정한 적이 없는 경우 **역량탭**을 클릭 시 **역량 상세 페이지**가 아니라 **최초 역량 설정 페이지**로 이동한다.
- 역량을 설정한 이력이 있는 경우 **역량탭**을 클릭 시 가장 최근에 설정된 **역량 상세 페이지**로 이동한다.

![image](https://user-images.githubusercontent.com/46308949/143985370-90854acf-e0f6-498a-bc5d-1ecb3cfbab3f.png)

### 최초 역량 설정하기
- 역량 설정이 없는 상황에서 역량탭 클릭 시 최초 역량 설정 페이지로 이동한다.
- 최초 역량 설정 페이지에서 프론트엔드 / 백엔드 기본 역량을 선택하여 역량을 설정할 수 있다.
- (제 3의 길) 역량 추가를 통해서도 역량을 추가할 수 있다.
- 최초 역량을 설정 할 경우 역량 상세 페이지로 이동한다. 새로운 역량이 설정되었기 때문에 역량 이력이 하나 추가된다.

##### 최초 역량 설정 API
```
POST /abilities/template/{template} 
```
- template은 `be` or `fe` 중 하나
- body없음

![image](https://user-images.githubusercontent.com/46308949/143983699-f9a8d054-df83-494e-9f22-e94f8144f1d5.png)

### 역량 관리하기
- 한번 이상 역량을 저장하면 역량탭 클릭시 마지막으로 설정한 **역량 상세 페이지**로 이동한다.
- 역량 상세 페이지에서 기존에 설정한 역량을 수정 / 삭제할 수 있으며 신규로 역량을 추가할 수 있다.
- 역량은 계층 구조를 가질 수 있으며 단계는 2단계 까지 가능하다.
- 역량 이력보기 버튼을 통해 역량 이력 목록을 조회할 수 있다.
- 변경 후 수정 버튼 클릭 시 새로운 역량 이력이 생긴다. (추후에 나올 역량 - 학습로그 맵핑 수정시에도 새로운 역량 이력이 생성됨)

![image](https://user-images.githubusercontent.com/46308949/143983806-b2699480-7ca8-4ae7-8d1e-ddcec7c8e37d.png)

### 역량에 학습로그 맵핑하기
- 역량 상세 페이지에서 역량에 학습로그를 맵핑 하기 위해서 학습로그를 먼저 조회한다.
- 선택한 학습로그 조회 결과가 상세 페이지 하단에 학습로그 목록에 노출된다.
- 학습로그 목록에서 맵핑하고자하는 역량을 학습로그 열에서 찾아서 등록한다.
- 조회해서 학습로그가 목록에 노출되더라도 맵핑을 하지 않으면 다음번 상세 페이지 조회 시 노출되지 않는다.

![image](https://user-images.githubusercontent.com/46308949/143984726-da4721cc-3868-45a5-801b-880230d312a6.png)

### 역량 - 학습로그 맵핑 관리하기
- 역량 페이지 조회 시 역량에 맵핑된 학습로그 목록이 노출된다.(맵핑이 안된 학습로그는 노출되지 않음)
- 기존에 맵핑된 내용을 제거할 수 있다. (제거되어도 목록에는 유지되며 화면을 리로드 할 경우 맵핑이 되지 않은 학습로그는 사라진다)
- 기존에 역량이 맵핑된 학습로그에 추가로 다른 역량을 맵핑할 수 있다.

### 역량 - 학습로그가 맵핑된 상태에서 역량 관리하기
- 학습로그가 맵핑된 역량을 수정할 경우 그대로 맵핑되어있음
- 학습로그가 맵핑된 역량을 삭제할 경우 맵핑된 학습로그는 해제됨

# 리포트

### 리포트 신규 생성
- 리포트탭 클릭 시 리포트 목록을 조회한다. (리포트를 생성한 적이 없으면 목록은 비어있음)
- 리포트를 작성하기 위해 **리포트 작성** 버튼을 누르면 리포트 작성 페이지로 이동한다.
- 리포트 작성 페이지에서 제목과 설명을 입력할 수 있으며 역량에 대한 가중치를 입력할 수 있다.
  - 이 때 가중치의 의미는 지난 리포트 작성 시점 이후부터 현재까지의 학습 중 어느 분야에 조금 더 무게를 두어 학습했는지에 대한 가중치이다.
- 차트는 [레이더 차트](https://www.chartjs.org/docs/latest/charts/radar.html)를 활용한다.

![image](https://user-images.githubusercontent.com/46308949/143987115-6f710e49-eafd-4334-9c34-0fbe55dc0134.png)

### 리포트 목록 조회
- 리포트를 작성한 이력이 있다면 리포트탭을 누를 경우 리포트 목록을 조회할 수 있다.
- 리포트 목록에는 [레이더 차트](https://www.chartjs.org/docs/latest/charts/radar.html)를 노출시킨다.

![image](https://user-images.githubusercontent.com/46308949/143987248-8409e808-d6ee-4047-b91f-b654a489a423.png)

# 기능 구현 진행 단계
## 1. 역량 및 역량 이력 기능 구현
- 최초 역량을 설정할 수 있다.
- 기존 역량에서 수정 / 삭제 또는 신규 역량 추가를 할 수 있다.
- 역량 수정(혹은 저장) 시 역량 이력이 하나씩 생성된다.
- 역량 이력 조회 버튼 클릭 시 역량 이력을 조회할 수 있다.

## 2. 역량 - 학습로그 맵핑 기능 구현
- 역량에 학습로그를 맵핑할 수 있고 제거할 수 있다.
- 역량 - 학습로그 맵핑 정보가 변경된 이후 저장되어도 역량 이력이 생성된다.

## 3. 맵핑된 역량 관리 기능 구현
- 역량이 삭제될 경우 맵핑된 학습로그가 해제된다.

## 4. 리포트 기능 구현
- 리포트 작성 시 역량 이력을 선택하여 리포트 정보로 활용한다.
- 리포트 작성 시 제목과 설명, 맵핑된 역량의 가중치를 입력한다.
- 리포트 저장을 한 후 리포트 목록을 조회할 수 있다.

**Comment by sunhpark42** [comment]
- 역량 히스토리를 선택해서 리포트 작성 => 최신 역량 히스토리에 대해서만 리포트 작성

**Reply by devhyun637** [reply]
+리포트에는 역량에 등록된 학습로그 노출(선택 가능)

**Comment by devhyun637** [comment]
(회의 이후)

역량탭에서 역량히스토리 버튼을 통해서 이전의 역량 볼 수 있음
